File: LegacySource/settlement/room/main/construction/ConstructionInit.cs
using System;
using settlement.room.main.construction;
using settlement.main;
using settlement.room.main;
using settlement.room.main.furnisher;
using settlement.room.main.util;
using settlement.tilemap.terrain;
using snake2d.util.datatypes;

namespace settlement.room.main.construction
{
    public class ConstructionInit
    {
        public readonly int Upgrade;
        public readonly Furnisher B;
        public readonly TBuilding Structure;
        public readonly int Degrade;
        public readonly RoomState State;

        private static RoomAreaWrapper Wrap = new RoomAreaWrapper();

        public ConstructionInit(int upgrade, Furnisher b, TBuilding structure, int degrade, RoomState state)
        {
            this.Upgrade = upgrade;
            this.Structure = structure;
            this.Degrade = degrade;
            this.State = state;
            this.B = b;
        }

        public ConstructionInit(Room room, int rx, int ry, bool broken)
        {
            this(room, rx, ry, room.Degrader(rx, ry) == null ? 0 : room.Degrader(rx, ry).GetData(), broken);
        }

        private ConstructionInit(Room room, int rx, int ry, int degrade, bool broken)
        {
            this(room.Upgrade(rx, ry), room.Constructor(), FindStructure(rx, ry), degrade, room.MakeState(rx, ry, broken));
        }

        public static TBuilding FindStructure(int rx, int ry)
        {
            Room room = SETT.ROOMS().Map.Get(rx, ry);

            if (room == null)
                return null;

            if (room is ConstructionInstance)
                return ((ConstructionInstance)room).Structure();

            ROOMA a = Wrap.Init(room, rx, ry);
            Wrap.Done();
            foreach (COORDINATE c in a.Body())
            {
                if (a.Is(c))
                {
                    TerrainTile t = TERRAIN().Get(c.X(), c.Y());
                    if (t is BuildingComponent)
                        return ((BuildingComponent)TERRAIN().Get(c.X(), c.Y())).Building();
                }
            }

            return TERRAIN().BUILDINGS.MUD;
        }
    }
}

File: settlement/room/main/construction/ConstructionInit.java
package settlement.room.main.construction;

import static settlement.main.SETT.TERRAIN;

import settlement.main.SETT;
import settlement.room.main.ROOMA;
import settlement.room.main.Room;
import settlement.room.main.furnisher.Furnisher;
import settlement.room.main.util.RoomAreaWrapper;
import settlement.room.main.util.RoomState;
import settlement.tilemap.terrain.TBuilding;
import settlement.tilemap.terrain.TBuilding.BuildingComponent;
import settlement.tilemap.terrain.Terrain.TerrainTile;
import snake2d.util.datatypes.COORDINATE;

public class ConstructionInit {

	public final int upgrade;
	public final Furnisher b;
	public final TBuilding structure;
	public final int degrade;
	public final RoomState state;
	
	private static RoomAreaWrapper wrap = new RoomAreaWrapper();
	
	public ConstructionInit(int upgrade, Furnisher b, TBuilding structure, int degrade, RoomState state) {
		this.upgrade = upgrade;
		this.structure = structure;
		this.degrade = degrade;
		this.state = state;
		this.b = b;		
	}
	
	public ConstructionInit(Room room, int rx, int ry, boolean broken) {
		this(room, rx, ry, room.degrader(rx, ry) == null ? 0 : room.degrader(rx, ry).getData(), broken);
	}
	
	private ConstructionInit(Room room, int rx, int ry, int degrade, boolean broken) {
		this(room.upgrade(rx, ry), room.constructor(), findStructure(rx, ry), degrade, room.makeState(rx, ry, broken));
	}
	
	public static TBuilding findStructure(int rx, int ry) {
		
		Room room = SETT.ROOMS().map.get(rx, ry);
		
		if (room == null)
			return null;
		
		if (room instanceof ConstructionInstance)
			return ((ConstructionInstance) room).structure();
		
		ROOMA a = wrap.init(room, rx, ry);
		wrap.done();
		for (COORDINATE c : a.body()) {
			if (a.is(c)) {
				TerrainTile t = TERRAIN().get(c.x(), c.y());
				if (t instanceof BuildingComponent)
					return ((BuildingComponent)TERRAIN().get(c.x(), c.y())).building();
			}
		}
		
		return TERRAIN().BUILDINGS.MUD;
	}
	
}
